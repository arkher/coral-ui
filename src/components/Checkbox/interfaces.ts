export type CheckboxProps = {
  label: string;
  value: boolean;
  onChange: () => void;
  disabled?: boolean;
  required?: boolean;
};
