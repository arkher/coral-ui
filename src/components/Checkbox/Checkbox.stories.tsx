import React from 'react';
import { withDesign } from 'storybook-addon-designs';
import Checkbox from './Checkbox';

export default {
  title: 'Componente/Checkbox',
  component: Checkbox,
  argTypes: {
    label: {
      control: { type: 'text' },
    },
    value: {
      control: { type: 'boolean' },
    },
    labelStyle: {
      control: { type: 'none' },
    },
    required: {
      control: { type: 'none' },
    },
    disabled: {
      control: { type: 'none' },
    },
  },
  decorators: [withDesign],
};

const urlHandoff =
  'https://www.figma.com/file/3raVfIADTUZCzFOOaQ9PMQ/HANDOFF-%7C-Core-Components-Mobile-%7C-Institucional?node-id=203%3A1636';

// Stories
export const Basic = ({ label, value }: any): React.ReactNode => (
  <Checkbox label={label} onChange={() => undefined} value={value} />
);

Basic.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};

export const Disabled = ({ label, value }: any): React.ReactNode => (
  <Checkbox label={label} onChange={() => undefined} value={value} disabled />
);

Disabled.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};

export const Required = ({ label, value }: any): React.ReactNode => (
  <Checkbox label={label} onChange={() => undefined} value={value} required />
);

Required.parameters = {
  design: {
    type: 'figma',
    url: urlHandoff,
  },
};
