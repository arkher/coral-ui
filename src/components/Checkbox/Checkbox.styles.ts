import { useTheme } from '@shopify/restyle';
import { StyleSheet } from 'react-native';
import { Theme } from '../..';

interface Props {
  checked: boolean;
  disabled: boolean;
  required: boolean;
}

const useStyles = ({ checked, disabled, required }: Props) => {
  const theme = useTheme<Theme>();

  let borderColor;
  if (disabled) borderColor = theme.colors.neutralLight;
  else if (required) borderColor = theme.colors.feedbackErrorBase;
  else borderColor = theme.colors.neutralBase;

  let bg;
  if (disabled) bg = theme.colors.neutralLight;
  else if (checked) bg = theme.colors.feedbackInfoBase;
  else bg = theme.colors.white;

  return StyleSheet.create({
    checkBox: {
      width: '100%',
      height: '100%',
      borderWidth: checked ? 0 : 2,
      borderRadius: 4,
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: bg,
      borderColor,
    },
  });
};

export default useStyles;
