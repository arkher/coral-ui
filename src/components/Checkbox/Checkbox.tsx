import React from 'react';
import { TouchableHighlight } from 'react-native';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import Box from '../Box';
import Text from '../Text';
import useStyles from './Checkbox.styles';
import { CheckboxProps } from './interfaces';

const Checkbox: React.FC<CheckboxProps> = (props: CheckboxProps) => {
  const { label, onChange, value, disabled, required } = props;

  const styles = useStyles({
    checked: value,
    disabled: !!disabled,
    required: !!required,
  });

  const handleChange = () => {
    if (onChange) {
      onChange();
    }
  };

  return (
    <Box flexDirection="row" alignItems="center">
      <TouchableHighlight
        style={styles.checkBox}
        onPress={handleChange}
        disabled={disabled}
      >
        {value ? <Icon name="check" size={16} color="#fff" /> : <Box />}
      </TouchableHighlight>

      <Text
        fs="xxxs"
        color={disabled ? 'neutralBase' : 'neutralDarkest'}
        ml={{ phone: 'nano', tablet: 'nano' }}
      >
        {label}
      </Text>
    </Box>
  );
};

export default Checkbox;
