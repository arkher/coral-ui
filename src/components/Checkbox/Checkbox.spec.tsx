import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react-native';
import { ThemeProvider } from '@shopify/restyle';
import Checkbox from './Checkbox';
import themeMaestro from '../../themes/maestro';

afterEach(cleanup);

jest.mock('react-native-vector-icons/MaterialCommunityIcons', () => 'Icon');

it('should be marked', () => {
  let value = false;
  const { getByTestId } = render(
    <ThemeProvider theme={themeMaestro}>
      <Checkbox
        label="Label da checkbox"
        value={value}
        onChange={() => {
          value = true;
        }}
      />
    </ThemeProvider>,
  );

  fireEvent.press(getByTestId('checkbox'));
  expect(value).toBe(true);
});

it('should have label', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={themeMaestro}>
      <Checkbox label="Label da checkbox" value onChange={() => undefined} />
    </ThemeProvider>,
  );

  expect(getByTestId('checkbox-label').props.children).toBe(
    'Label da checkbox',
  );
});

test('should Checkbox render required feedback', async () => {
  const { toJSON } = render(
    <ThemeProvider theme={themeMaestro}>
      <Checkbox
        label="Label da checkbox"
        required
        value
        onChange={() => undefined}
      />
    </ThemeProvider>,
  );

  expect(toJSON()).toMatchSnapshot();
});

test('should Checkbox render correctly', async () => {
  const { toJSON } = render(
    <ThemeProvider theme={themeMaestro}>
      <Checkbox label="Label da checkbox" value onChange={() => undefined} />
    </ThemeProvider>,
  );

  expect(toJSON()).toMatchSnapshot();
});
