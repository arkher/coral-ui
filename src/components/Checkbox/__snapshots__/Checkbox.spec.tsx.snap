// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should Checkbox render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
      },
      Array [
        Object {},
      ],
    ]
  }
  testID="Box"
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#106CEE",
        "borderColor": "#B4BBC2",
        "borderRadius": 4,
        "borderWidth": 0,
        "height": 24,
        "justifyContent": "center",
        "width": 24,
      }
    }
    testID="checkbox"
  >
    <Icon
      color="#fff"
      name="check"
      size={16}
    />
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#2D3338",
          "display": "flex",
          "fontFamily": "\\"Inter\\", sans-serif",
          "fontWeight": "400",
          "marginLeft": 8,
        },
        Array [
          Object {
            "fontSize": 14,
          },
        ],
      ]
    }
    testID="checkbox-label"
  >
    Label da checkbox
  </Text>
</View>
`;
